### ISSUE TITLE
2759 - RequiredByKeys

### ISSUE THREAD

**Issue by blinpete** [type-challenges/type-challenges#10826]
## TL;DR

```ts
// constrained R
type RequireKeys<T, R extends keyof T = keyof T> = T & { [P in R]-?: T[P] }

// free R
type RequiredByKeys<T, R = keyof T> = RequireKeys<T, R & keyof T>
```

## About the signature

It makes more sense to have a constrained generic type `R` for two reasons:
1. you get autocompletion when typing the keys to be required
2. you get a static error when trying to require a key which is out of `keyof T`

So, the signature should be `<T, R extends keyof T = keyof T>`.

And for those cases when you really have to require the keys that possibly don't exist in `T` you can easily set the `R` generic free just wrapping up the `RequireKeys` as shown in TLDR.

## Some explanation

The solution itself is pretty strightforward, it makes "Mapped type" with required keys and then intersects it with the original type.

The only thing worth mentioning is that the order matters in `T & { [P in R]-?: T[P] }`. Even though intersection is commutative in terms of keys, in terms of key modifiers it is not.

## About wrappers (for those who are new to typescript)

```ts
type Wrap<T> = { [K in keyof T]: T[K] }
type Wrap<T> = Omit<T, never>
type Wrap<T> = Pick<T, keyof T>
```
All those wrappers (called `Merge`, `Copy`, `MapObj`, etc etc) you can see in other answers are completely redundant. You don't have to use them in your real-world code. They are just to pass unit tests in that [Typescript playground](https://www.typescriptlang.org/play).

## Tags

`#mapped-type` `#intersection`
